pkg/provisioning/provisioning.go

```diff
@@ -0,0 +1,98 @@
+package provisioning
+
+import (
+	metalkubev1alpha1 "github.com/metalkube/baremetal-operator/pkg/apis/metalkube/v1alpha1"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("provisioner")
+
+/*
+Package provisioning defines the API for talking to the provisioning backend.
+*/
+
+// Provisioner holds the state information for talking to the
+// provisioning backend.
+//
+// NOTE(dhellmann): Provisioner will eventually become an interface,
+// but not until we need to have more than one backend.
+type Provisioner struct {
+}
+
+// Register the host with Ironic.
+func (p *Provisioner) ensureExists(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
+	reqLogger := log.WithValues("host", host.Name)
+	if host.Status.Provisioning.ID == "" {
+		host.Status.Provisioning.ID = "temporary-fake-id"
+		reqLogger.Info("setting provisioning id",
+			"provisioningID", host.Status.Provisioning.ID)
+		dirty = true
+	}
+	return dirty, nil
+}
+
+// ValidateManagementAccess tests the connection information for the
+// host to verify that the location and credentials work.
+func (p *Provisioner) ValidateManagementAccess(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
+	reqLogger := log.WithValues("host", host.Name)
+	reqLogger.Info("testing management access")
+	if dirty, err := p.ensureExists(host); err != nil {
+		return dirty, err
+	}
+	return dirty, nil
+}
+
+// InspectHardware updates the HardwareDetails field of the host with
+// details of devices discovered on the hardware. It may be called
+// multiple times, and should return true for its dirty flag until the
+// inspection is completed.
+func (p *Provisioner) InspectHardware(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
+	reqLogger := log.WithValues("host", host.Name)
+	reqLogger.Info("inspecting hardware", "status", host.OperationalStatus())
+
+	if dirty, err = p.ensureExists(host); err != nil {
+		return dirty, err
+	}
+
+	if host.OperationalStatus() != metalkubev1alpha1.OperationalStatusInspecting {
+		// The inspection just started.
+		reqLogger.Info("starting inspection by setting state")
+		host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusInspecting)
+		return true, nil
+	}
+
+	// The inspection is ongoing. We'll need to check the ironic
+	// status for the server here until it is ready for us to get the
+	// inspection details. Simulate that for now by creating the
+	// hardware details struct as part of a second pass.
+	if host.Status.HardwareDetails == nil {
+		reqLogger.Info("continuing inspection by setting details")
+		host.Status.HardwareDetails = &metalkubev1alpha1.HardwareDetails{}
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Deprovision cleans the host and removes it from the external
```

Comment:
I noticed earlier that you named the function `Remove`, and I see here from the comment that I assume you also considered `Deprovision`. I sort of like `Deprovision` when applying the verb to a host, but could go either way. I guess `provisioner.Remove()` implies that the provisioner is removing whatever it had previously added? 

In any case, either name is fine, but this comment should match.

Reply:
IMO stating that this function "removes it from the external provisioning tool" reveals too much about the implementation details of a particular provisioner. I suggest we stick to describing what changes will be made to the host itself when calling each function.

Reply:
I waffled on the name. I'll switch it back to Deprovision().